// discord.js-commando/src/commands/util/unknown-command.js
const Discord = require('discord.js');
const Command = require('../base');

const path = require('path');
const config = require(path.join(__dirname, '..', '..', '..', '..', '..', 'config', 'config.json'));

module.exports = class UnknownCommandCommand extends Command {
	constructor(client) {
		super(client, {
			name: 'unknown-command',
			group: 'util',
			memberName: 'unknown-command',
			description: 'Displays help information for when an unknown command is used.',
			examples: ['unknown-command kickeverybodyever'],
			unknown: false, // true
			hidden: true
		});
	}

	run(msg) {
		const unknownCommandMessage = new Discord.MessageEmbed()
            .setColor(config.embedSettings.color)
            .setAuthor('Unknown command', `https://images.emojiterra.com/mozilla/512px/26a0.png`, '')
			.setDescription(`Use ${msg.anyUsage('help',msg.guild ? undefined : null,msg.guild ? undefined : null)} to view the command list.`)

            //.setTimestamp()
			.setFooter(config.embedSettings.footer, config.embedSettings.footer_url);
		
		msg.channel.send(unknownCommandMessage);

		/*return msg.reply(
			`Unknown command. Use ${msg.anyUsage(
				'help',
				msg.guild ? undefined : null,
				msg.guild ? undefined : null
			)} to view the command list.`
		);*/
	}
};
