// discord.js-commando/src/commands/commands/enable.js
const Discord = require('discord.js');
const { oneLine } = require('common-tags');
const Command = require('../base');
const pool = require('../../../../../lib/dbpool');

const path = require('path');
const { deleteMySQLRecord } = require('discord.js-commando/src/commands/util/prefix');
const { embedSettings } = require(path.join(__dirname, '..', '..', '..', '..', '..', 'config', 'config.json'));

module.exports = class EnableCommand extends Command {
	constructor(client) {
		super(client, {
			name: 'enable',
			aliases: ['enable-command', 'cmd-on', 'command-on'],
			group: 'commands',
			memberName: 'enable',
			description: 'Enables a command or command group.',
			details: oneLine`
				The argument must be the name/ID (partial or whole) of a command or command group.
				Only administrators may use this command.
			`,
			examples: ['enable util', 'enable Utility', 'enable prefix'],
			guarded: true,

			args: [
				{
					key: 'cmdOrGrp',
					label: 'command/group',
					prompt: 'Which command or group would you like to enable?',
					type: 'group|command'
				}
			]
		});
	}

	hasPermission(msg) {
		if(!msg.guild) return this.client.isOwner(msg.author);
		return this.client.isOwner(msg.author) || msg.member.hasPermission('ADMINISTRATOR');
	}

	run(msg, args) {
		const group = args.cmdOrGrp.group;
		if(args.cmdOrGrp.isEnabledIn(msg.guild, true)) {
			return msg.reply(
				`The \`${args.cmdOrGrp.name}\` ${args.cmdOrGrp.group ? 'command' : 'group'} is already enabled${
					group && !group.isEnabledIn(msg.guild) ?
					`, but the \`${group.name}\` group is disabled, so it still can't be used` :
					''
				}.`
			);
		}
		args.cmdOrGrp.setEnabledIn(msg.guild, true);
		if (!group && msg.guild) module.exports.updateMySQLRecord(msg.guild.id, args.cmdOrGrp.id);
		return msg.reply(
			`Enabled the \`${args.cmdOrGrp.name}\` ${group ? 'command' : 'group'}${
				group && !group.isEnabledIn(msg.guild) ?
				`, but the \`${group.name}\` group is disabled, so it still can't be used` :
				''
			}.`
		);
	}

	static updateMySQLRecord(guildID, module) {
		pool.query(`SELECT * FROM modules WHERE guildID = '${guildID}'`, function (err, result) {
			if (err) throw err;
			//console.log(result);

			let modulesRecord = {};
			if (result.length != 0) modulesRecord = JSON.parse(result[0].modules);

			if (modulesRecord[module] == false) {
				// Deletes record and stops if all modules will be enabled
				if (Object.keys(modulesRecord).length == 1) {
					var deleteRecord = `DELETE FROM modules WHERE guildID = '${guildID}'`;
					pool.query(deleteRecord, function (err, result) {
						if (err) throw err;
						//console.log("Number of records deleted: " + result.affectedRows);
					});
				}

				delete modulesRecord[module]; // Deletes module from record (not in record defaults to enabled)
				var changeModuleStatus = `UPDATE modules SET modules = '${JSON.stringify(modulesRecord)}' WHERE guildID = '${guildID}'`;
				pool.query(changeModuleStatus, function (err, result) {
					if (err) {throw err;}
					//console.log(result.affectedRows + " record(s) updated");
				});
			}
		});
	}
};
